plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,238 @@
+	"time"
+)
+
+type Parser struct {
+	MetricName      string   `toml:"metric_name"`
+	SchemaRegistry  string   `toml:"avro_schema_registry"`
+	Schema          string   `toml:"avro_schema"`
+	Measurement     string   `toml:"avro_measurement"`
+	Tags            []string `toml:"avro_tags"`
+	Fields          []string `toml:"avro_fields"`
+	Timestamp       string   `toml:"avro_timestamp"`
+	TimestampFormat string   `toml:"avro_timestamp_format"`
+	DiscardArrays   bool     `toml:"avro_discard_arrays"`
+	FieldSeparator  string   `toml:"avro_field_separator"`
+	DefaultTags     map[string]string
+	TimeFunc        func() time.Time
+
+	Log telegraf.Logger `toml:"-"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var schema string
+
+	if len(buf) < 5 {
+		err := fmt.Errorf("buf is %d bytes; must be at least 5", len(buf))
+		return nil, err
+	}
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
+
+	switch {
+	case p.SchemaRegistry != "":
+		schemaRegistry := NewSchemaRegistry(p.SchemaRegistry)
+		retrSchema, err := schemaRegistry.getSchema(schemaID)
+		if err != nil {
+			return nil, err
+		}
+		schema = retrSchema
+	default:
+		schema = p.Schema
+	}
+
+	codec, err := goavro.NewCodec(schema)
+
+	if err != nil {
+		return nil, err
+	}
+
+	native, _, err := codec.NativeFromBinary(binaryData)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := p.createMetric(native, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{m}, nil
+}
+
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	metrics, err := p.Parse([]byte(line))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(metrics) != 1 {
+		return nil, fmt.Errorf("line contains multiple metrics")
+	}
+
+	return metrics[0], nil
+}
+
+func (p *Parser) SetDefaultTags(tags map[string]string) {
+	p.DefaultTags = tags
+}
+
+func (p *Parser) parseTimestamp(timestamp interface{}) (time.Time, error) {
+	if timestamp == nil {
+		return p.TimeFunc(), nil
+	}
+
+	if p.TimestampFormat == "" {
+		return p.TimeFunc(), fmt.Errorf("must specify timestamp format")
+	}
+
+	metricTime, err := internal.ParseTimestamp(p.TimestampFormat, timestamp, "UTC")
+	if err != nil {
+		return p.TimeFunc(), err
+	}
+
+	return metricTime, nil
+}
+
+func (p *Parser) createMetric(native interface{}, schema string) (telegraf.Metric, error) {
+	deep, ok := native.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("cannot cast native interface {} to map[string]interface{}")
+	}
+
+	metricTime, err := p.parseTimestamp(nestedValue(deep[p.Timestamp]))
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+
+	for _, tag := range p.Tags {
+		if value, ok := deep[tag]; ok {
+			tags[tag] = fmt.Sprintf("%v", nestedValue(value))
+		} else {
+			p.Log.Infof("tag %s value was %v; not added to tags", tag, value)
+		}
```

Comment:
This rather sounds like a debug message. Otherwise you may risk log-spamming...

Reply:
More likely I should make it a warning.  If we fail to unpack the value, something weird is going on.